crates/matrix-sdk-crypto/src/session_manager/sessions.rs

```diff
@@ -344,6 +345,37 @@ impl SessionManager {
         self.failures.extend(failed_servers);
         self.failures.remove(successful_servers);
 
+        // build a map of user_id -> device_id -> key for each device to we can start
+        // a session with.
+        let mut device_map: BTreeMap<
```

Comment:
I don't understand the motivation here. Why do we want to treat this single session creation error in this special way?

Furthermore, if we really want to treat this specially why don't we first collect the failures and then acquire the lock only once, and only emit a single log line. As this is written I don't think it offers any advantage over the previous way of handling those failures. 

I would probably use `filter_map()` or [`partition_map()`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.partition_map) from itertools for this.

Reply:
> I don't understand the motivation here. Why do we want to treat this single session creation error in this special way?

Well basically this is groundwork for https://github.com/matrix-org/matrix-rust-sdk/pull/2805. Once that lands, we are trying to correlate the request with the response, which means that we discover this "missing key" error rather earlier than the other types of errors.

> Furthermore, if we really want to treat this specially why don't we first collect the failures and then acquire the lock only once, and only emit a single log line.

Yup, this is a good suggestion, thanks! Will make this change.

> I would probably use `filter_map()` or [`partition_map()`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.partition_map) from itertools for this.

Again, going back to #2805, this won't be as easy, because we're trying to correlate two sets of maps. I think we'd end up undoing such a change.